Add getCodec method in ForgeRegistry

Frinn38 opened the pull request:
This is mostly the same PR as #7190 a bit rewritten and simplified.
The main change is that this PR add a getCodec() method  in IForgeRegistry instead of having the registry implements the whole Codec interface (asked by LexManos on discord in modder-support-116, 03/08/2020)

This PR add a getCodec() method in IForgeRegistry interface and implement it's methods encode & decode in ForgeRegistry class.
It also add a test mod to show a possible usage of this registry codec and prove that this implementation of encode and decode works.

The Codec returned by the method will be a useful tool to serialize/deserialize registry entries in Json or NBT.
It is compatible with all Forge Registries including custom ones.

Frinn38 committed:
add getCodec() to IForgeRegistry and testmod

Added a getCodec() method in IForgeRegistry interface and implement it's methods encode & decode in ForgeRegistry class.
Add a test mod to show a possible usage of this registry codec and prove that this implementation of encode and decode works.

kashike commented on src/main/java/net/minecraftforge/registries/ForgeRegistry.java:
```diff
@@ -101,6 +106,42 @@
     private final RegistryKey<Registry<V>> key;
     private final RegistryBuilder<V> builder;
+    
+    private final Codec<V> CODEC = new Codec<V>() {
```
This is not a `static` field.
```suggestion
    private final Codec<V> codec = new Codec<V>() {
```

Frinn38 replied:
Yes, changed it to lowercase.

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/ForgeRegistry.java:
```diff
@@ -101,6 +106,42 @@
+            {
+                return DataResult.error("Unknown registry element " + input);
+            }
+            T toMerge = ops.compressMaps() ? ops.createInt(getID(input)) : ops.createString(key.toString());
```
Question: why is the `ResourceLocation` converted to a string via `.toString()` instead of its `Codec` (from `ResourceLocation.CODEC`)? (aside from this being a formatted copy from vanilla's `Registry#encode`)
(Yes, I understand this works as vanilla uses it, but I'm curious as to why not the codec too; disclaimer: I have only a passing familiarity with codecs)

Frinn38 replied:
As you sait it's a copy of the vanilla registry implementation. But just look at the ResourceLocation.CODEC and you will see that it's encode method is basically ResourceLocation::toString.
So there is no need of the ResoureceLocation codec in this case.
Also the codec would return a DataResult<T> with T being the String so I would need to add severals checks to handle the case the codec failed to encode the ResourceLocation, this should not happen because the encode is just the toString() method but still the code for the checks would be needed to retrieve the encoded String from the DataResult.

Frinn38 committed:
Changed codec Field to lowercase

Frinn38 committed:
Merge branch '1.16.x' into 1.16.x

Frinn38 committed:
Change anonymous codec to inner RegistryCodec class

pupnewfster commented on src/main/java/net/minecraftforge/registries/ForgeRegistry.java:
```diff
@@ -839,6 +850,42 @@ public Snapshot makeSnapshot()
         return ret;
     }
 
+    public class RegistryCodec implements Codec<V> {
```
Is there a reason this implementation detail is public instead of private?

Frinn38 replied:
Not at all, it can (and should probably) be private.

Frinn38 replied:
done

Frinn38 committed:
Change RegistryCodec inner class to private

Frinn38 committed:
Merge branch '1.16.x' into 1.16.x

Frinn38 committed:
Update Test Mod
Add a test for Compressed Ops (use integer ids instead of ResourceLocation ones)
Add some comments

Frinn38 committed:
Merge remote-tracking branch 'origin/1.16.x' into 1.16.x

TheCurle commented:
Some preliminary testing shows it works. Codecs are a known nightmare, so i'll test it more thoroughly before recommending anything.

Your test class is missing a license header. That seems to be the only thing holding this back for the moment, though.

Frinn38 committed:
Add license header

Frinn38 commented:
Completely forgot the License header, now fixed.
I agree that Codecs can be a pain (at least they are relatively easy to use but still a nightmare to understand) but I believe that in the future Mojang will use them more and more and modders will have to follow, and when that time come, having a codec for all ForgeRegistryEntry will be necessary (it's just my guess, maybe Mojang will just void them and turn back to good old serializers but I doubt)

Frinn38 committed:
Merge branch '1.16.x' into 1.16.x

Frinn38 committed:
Merge branch '1.16.x' into 1.16.x

ChampionAsh5357 commented on src/main/java/net/minecraftforge/registries/ForgeRegistry.java:
```diff
@@ -185,6 +191,11 @@ public String getTagFolder()
     }
 
+    @Override
+    public Codec<V> getCodec() {
```
Bracket on new line.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/registries/ForgeRegistry.java:
```diff
@@ -839,6 +850,42 @@ public Snapshot makeSnapshot()
         return ret;
     }
 
+    private class RegistryCodec implements Codec<V> {
```
Bracket on new line.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/registries/ForgeRegistry.java:
```diff
@@ -839,6 +850,42 @@ public Snapshot makeSnapshot()
+        }
+
+        @Override
+        public <T> DataResult<T> encode(V input, DynamicOps<T> ops, T prefix) {
```
Bracket on new line.

ChampionAsh5357 commented on src/test/java/net/minecraftforge/debug/misc/CodecRegistryTest.java:
```diff
@@ -0,0 +1,86 @@
+import org.apache.logging.log4j.Logger;
+
+@Mod("codec_registry_test")
+public class CodecRegistryTest {
```
Bracket on new line.

ChampionAsh5357 commented on src/test/java/net/minecraftforge/debug/misc/CodecRegistryTest.java:
```diff
@@ -0,0 +1,86 @@
+        ForgeRegistries.ITEMS.getCodec().fieldOf("item").forGetter(Pair::getSecond)
+    ).apply(codecInstance, Pair::of));
+
+    public CodecRegistryTest() {
```
Bracket on new line.

ChampionAsh5357 commented on src/test/java/net/minecraftforge/debug/misc/CodecRegistryTest.java:
```diff
@@ -0,0 +1,86 @@
+        FMLJavaModLoadingContext.get().getModEventBus().addListener(this::commonSetup);
+    }
+
+    public void commonSetup(final FMLCommonSetupEvent event) {
```
Bracket on new line.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/registries/ForgeRegistry.java:
```diff
@@ -839,6 +850,42 @@ public Snapshot makeSnapshot()
 
+    private class RegistryCodec implements Codec<V> {
+        @Override
+        public <T> DataResult<Pair<V, T>> decode(DynamicOps<T> ops, T input) {
```
Bracket on new line.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/registries/ForgeRegistry.java:
```diff
@@ -839,6 +850,42 @@ public Snapshot makeSnapshot()
+                    }
+                    V val = ForgeRegistry.this.getValue(id);
+                    return DataResult.success(val);
+                }).map(v -> Pair.of(v, input));
```
The `Pair` returned by `#decode` represents the data that was successfully decoded to the object and the data that failed to decode. This says that the output was both fully encoded and decoded like schrodinger's cat.

Frinn38 replied:
You're right, vanilla registry codec return ops.empty() here...

ChampionAsh5357 reviewed (changes requested):
Leaving some general fixes that's needed. From my testing, it seems to be fine besides from the misunderstanding involving `#decode`. Will wait on @TheCurle to confirm whenever she has the time. You should also update this to 1.17 since that is now latest.

Frinn38 committed:
Put open brackets on new line and fix decode returning the input object instead of empty.

Frinn38 committed:
Merge branch '1.16.x' into 1.16.x

Frinn38 commented:
I will add the same PR for 1.17 now that it's the latest version.

Frinn38 committed:
Update mods.toml

Frinn38 committed:
Merge branch '1.16.x' into 1.16.x

Frinn38 committed:
Merge branch '1.16.x' into 1.16.x

Frinn38 committed:
Merge branch '1.16.x' into 1.16.x

Frinn38 commented:
Closed as it's finally merged in 1.18

ChampionAsh5357 commented:
Why wouldn't this be an LTS PR for merging into 1.16?
